core/variant/array.cpp

```diff
@@ -54,16 +54,6 @@ void Array::_ref(const Array &p_from) const {
 
 	ERR_FAIL_COND(!_fp); // should NOT happen.
 
-	if (unlikely(_fp->read_only != nullptr)) {
-		// If p_from is a read-only array, just copy the contents to avoid further modification.
-		_unref();
-		_p = memnew(ArrayPrivate);
-		_p->refcount.init();
-		_p->array = _fp->array;
-		_p->typed = _fp->typed;
-		return;
-	}
-
```

Comment:
I believe this was done to allow mutability on assignments, e.g.:

```gdscript
const arr = [1, 2, 3]
var copy = arr # Create copy so it's a regular variable.
copy.push_back(4) # Only changes `copy`.
print(copy) # [1, 2, 3, 4]
print(arr) # [1, 2, 3]
```

Which wouldn't work with this change (`push_back()` would fail). But I don't think there's a valid use case for this and the user can call `duplicate()` if this is needed.